src/Nri/Ui/Mark/V6.elm

```diff
@@ -69,72 +74,100 @@ viewWithOverlaps :
-                in
-                ( Set.fromList maybeStringSorter (List.map .name marks)
-                , acc
-                    ++ (case List.filterMap .name before of
-                            [] ->
-                                [ segment startStyles ]
-
-                            names ->
-                                [ span [ css startStyles ]
-                                    [ viewInlineTag
-                                        [ Css.display Css.none
-                                        , MediaQuery.highContrastMode
-                                            [ Css.property "forced-color-adjust" "none"
-                                            , Css.display Css.inline |> Css.important
-                                            , Css.property "color" "initial" |> Css.important
-                                            ]
-                                        ]
-                                        (String.Extra.toSentenceOxford names)
-                                    ]
-                                , segment []
-                                ]
-                       )
-                )
+    overlappingStyles segments
+        |> List.concatMap
+            (\( content, maybeLabel, styles ) ->
+                case maybeLabel of
+                    Nothing ->
+                        [ viewSegment content styles ]
+
+                    Just label ->
+                        [ label, viewSegment content styles ]
             )
-            ( Set.empty maybeStringSorter, [] )
-        |> Tuple.second
 
 
-ignoreRepeats : Set (Maybe String) -> List Mark -> List Mark
-ignoreRepeats lastMarks list =
-    List.filter (\x -> not (Set.memberOf lastMarks x.name)) list
+{-| Compute the styles required to mark the segments.
+
+You can use this if you require more control over the structure of how marked elements are laid out than `viewWithOverlaps` provides.
+
+The `Maybe (Html msg)` result is a label element that should be rendered before the current segment.
+
+-}
+overlappingStyles : List ( content, List Mark ) -> List ( content, Maybe (Html msg), List Style )
+overlappingStyles segments =
+    let
+        -- We can't detect the end of a span of marks until after we are past it!
+        -- So we need to go back to the last set of styles and add the ended styles there
+        updateEndRevStyles patchRevStyles endedMarks =
+            case ( patchRevStyles, endedMarks ) of
+                ( [], _ ) ->
+                    patchRevStyles
+
+                ( _, [] ) ->
+                    patchRevStyles
+
+                ( ( prevContent, prevLabel, prevStyles ) :: otherPrevStyles, _ ) ->
+                    ( prevContent, prevLabel, prevStyles ++ (tagAfterContent endedMarks :: List.concatMap .endStyles endedMarks) ) :: otherPrevStyles
+
+        { priorMarks, revStyles } =
+            List.foldl
+                (\( content, marks ) state ->
```

Comment:
🎨 maaaaybe rename `marks` to `currentMarksList` or similar? might help understanding with so many collections of marks floating around.

Reply:
Done!